lib/bolt/applicator.rb

```diff
@@ -50,11 +52,23 @@ def compile(target, ast, plan_vars)
       JSON.parse(out)
     end
 
+    def validate_hiera_config(hiera_config)
+      if File.exist?(File.path(hiera_config))
+        data = File.open(File.path(hiera_config), "r:UTF-8") { |f| YAML.safe_load(f.read) }
+        unless data['version'] == 5
+          raise ApplyError.new("All Targets", "Hiera v5 is required.")
+        end
+        hiera_config
+      end
+    end
+
     def apply(args, apply_body, scope)
       raise(ArgumentError, 'apply requires a TargetSpec') if args.empty?
       type0 = Puppet.lookup(:pal_script_compiler).type('TargetSpec')
       Puppet::Pal.assert_type(type0, args[0], 'apply targets')
 
+      @hiera_config = validate_hiera_config(@hiera_config)
```

Comment:
Would it make sense to do the validation upon assignment in initialization?

Reply:
I think we should do it as early as possible.

Reply:
I think it makes sense to do when the Applicator is initialized. It may trigger if you're not using apply, but the hiera config is entirely opt-in, so if you're doing something with Puppet I think we should always say you have an invalid hiera config. Later we may use it for things besides apply.